### PR TITLE
mavros_extras: Ported vision_pose_estimate plugin for ROS2

### PR DESCRIPTION
**Pull request by Peter010103**
Currently work in progress, but would appreciate comments on the changes.

### CODE REVIEW

**Comment by Peter010103** [comment]
For reference, I get two warnings during compilation related to line 197 in `vision_pose_estimate.cpp`

Here is the exact error message :)

File: mavros_extras/mavros_plugins.xml [hunk 1]
```diff
@@ -21,5 +21,10 @@ to the FCU and/or receive landing target tracking data coming from the FCU.</des
     <description>@brief 3DR Radio plugin.
 @plugin tdr_radio</description>
   </class>
+
+  <class name="vision_pose" type="mavros::plugin::PluginFactoryTemplate&lt;mavros::extra_plugins::VisionPoseEstimatePlugin&gt;" base_class_type="mavros::plugin::PluginFactory">
+    <description>@brief Vision pose estimate plugin
+@plugin vision_pose</description>
+  </class>
```
Comment (vooon) [inline mavros_extras/mavros_plugins.xml:28]: That's generated code by cog. I'll may do that after merge.

File: mavros_extras/src/plugins/vision_pose_estimate.cpp [hunk 2]
```diff
@@ -15,57 +15,82 @@
...
-          " -> " << tf_child_frame_id);
-      tf2_start("VisionPoseTF", &VisionPoseEstimatePlugin::transform_cb);
-    } else {
-      vision_sub = sp_nh.subscribe("pose", 10, &VisionPoseEstimatePlugin::vision_cb, this);
-      vision_cov_sub = sp_nh.subscribe(
-        "pose_cov", 10, &VisionPoseEstimatePlugin::vision_cov_cb,
-        this);
-    }
+    node_declate_and_watch_parameter(
+      "tf/listen", false, [&](const rclcpp::Parameter & p) {
```
Comment (vooon) [inline mavros_extras/src/plugins/vision_pose_estimate.cpp:62]: Use dot to separate subelements, e.g. `tf.listen`.

File: mavros_extras/src/plugins/vision_pose_estimate.cpp [hunk 3]
```diff
@@ -15,57 +15,82 @@
...
+            tf2_start("VisionPoseTF", &VisionPoseEstimatePlugin::transform_cb);
+          } else{
+            vision_sub = node->create_subscription<geometry_msgs::msg::PoseStamped>(
+              "~/pose", 10, std::bind(
+                &VisionPoseEstimatePlugin::vision_cb, this, _1));
+            vision_cov_sub = node->create_subscription<geometry_msgs::msg::PoseWithCovarianceStamped>(
+              "~/pose_cov", 10, std::bind(
+                &VisionPoseEstimatePlugin::vision_cov_cb, this, _1));
+          }
+        });
```
Comment (vooon) [inline mavros_extras/src/plugins/vision_pose_estimate.cpp:78]: NOTE: place "switching" parameter after it's config parameters. Now you would try to start TF with unset frame ids.
Also it's good to make those parameters to be possible to change in operation (you'd have to reset all things to initial state, then init subs//pubs again).

File: mavros_extras/src/plugins/vision_pose_estimate.cpp [hunk 4]
```diff
@@ -111,18 +134,22 @@ class VisionPoseEstimatePlugin : public plugin::PluginBase,
...
 
-    vp.usec = stamp.toNSec() / 1000;
+    vp.usec = stamp.nanoseconds() / 1000;
     // [[[cog:
     // for f in "xyz":
     //     cog.outl("vp.%s = position.%s();" % (f, f))
     // for a, b in zip("xyz", ('roll', 'pitch', 'yaw')):
     //     cog.outl("vp.%s = rpy.%s();" % (b, a))
     // ]]]
+
```
Comment (vooon) [inline mavros_extras/src/plugins/vision_pose_estimate.cpp:152]: Please don't change cog-generated code.

**Review by vooon — COMMENTED** [review]
Overall looks good. Warning in FTF - i think you could replace index var types with Eigen one.

Also look carefully to tf mixin, it segfaults for landing_target.

**Comment by Peter010103** [comment]
Hey @vooon . I'm a little confused as to whether this plugin is actually working or not.

When I do `ros2 run mavros mavros_node --ros-args --remap __ns:=/fpvquad_0/mavros -p fcu_url:=/dev/serial1:921600` and then run `ros2 topic list`, I can definitely see and 'echo' the ROS2 topic `/fpvquad_0/mavros/vision_pose/pose`. I am running the mavros node on a Raspberry Pi onboard the drone, which is connected to the same domain as my PC where I provide and publish fake pose data to the above topic.

I wanted to check whether I am actually receiving these topics on the flight computer, so I tried 3 methods:
1. Log the data by adding the `vehicle_visual_odometry` uORB topic in `etc/logging/logger_topics.txt` on the SD card.
2. Go to MAVLink Console and check that the topic is listed in `/obj`
3. Go to MAVLink Console and do `uorb top`

For 1, when I check the .ulg log file (and convert it to csv) I don't see the `vehicle_visual_odometry` topic 
For 2, after running the mavros node I can see the `vehicle_visual_odometry` topic listed
For 3, I get the below message. (I assume the instance of 0 is saying the FC did not receive the mavlink messages)
Screenshot from 2021-08-31 12-57-34

Could you provide some pointers?

**Comment by vooon** [comment]
@Peter010103 that's sounds similar to #1601. But first you probably need to check that plugin is loaded.
Note that you can intercept all messages on UAS link (i.e. `/uas1/mavlink_sink`).

**Comment by Peter010103** [comment]
@vooon I could see the plugin was loaded and was able to see the messages by doing a ros2 topic echo of `/uas1/mavlink_sink` :)

As an extra measure, I connected the RX and TX of the Raspberry to a logic analyzer and verified that I was receiving MAVLink 2.0 messages with msg_id: 102 (the VISION_POSITION_ESTIMATE message)

File: mavros_extras/src/plugins/vision_pose_estimate.cpp (hunk 4 above)
Comment (dayjaby) [reply mavros_extras/src/plugins/vision_pose_estimate.cpp:152]: I think this was caused by my sanitize PR: https://github.com/mavlink/mavros/pull/1603
I didn't find a way to get rid of the warnings of the `\` in the end without creating that extra blank line. So all the blame on me :(

**Review by vooon — APPROVED** [review]
I think i'll merge that PR as is and then fix it.

**Pull request merged**